ass4.cpp

```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
+                counter++;
+            }
+        }
+    }
+    return counter;
+}
 
-void flood_fill_queued(vector<vector<bool> > &table, int y, int x) {
-    deque<pair<int, int>> tile_deque;
-    tile_deque.push_back({y,x});
+// Algorithm "Union-Find" using Tree to store disjoint components as unique islands
+
+class Disjoint_sets {
+    vector<int> rank;
+    vector<int> parent;
     
-    while (!tile_deque.empty()) {
-        pair<int, int> cur_tile = tile_deque.front();
+public:
+    Disjoint_sets(int n) {
+        rank.resize(n, 0);
+        for (int i = 0; i < n; ++i) {
+            parent.push_back(i);
+        }
+    }
     
-        tile_deque.pop_front();
+    int find(int x) {
+        if (parent[x] != x) {
+            return find(parent[x]);
+        }
+        return x;
+    }
+    
+    void func_union(int x, int y) {
+        int x_ancestor = find(x);
+        int y_ancestor = find(y);
         
-        if (cur_tile.first - 1 >= 0 && table[cur_tile.first - 1][cur_tile.second]) {
-            table[cur_tile.first - 1][cur_tile.second] = false;
-            tile_deque.push_back({cur_tile.first - 1, cur_tile.second});
+        if (x_ancestor == y_ancestor) {
+            return;
         }
-        if (cur_tile.first + 1 < table.size() && table[cur_tile.first + 1][cur_tile.second]) {
-            table[cur_tile.first + 1][cur_tile.second] = false;
-            tile_deque.push_back({cur_tile.first + 1, cur_tile.second});
+        
+        if (rank[x_ancestor] < rank[y_ancestor]) {
+            parent[x_ancestor] = y_ancestor;
         }
-        if (cur_tile.second - 1 >= 0 && table[cur_tile.first][cur_tile.second - 1]) {
-            table[cur_tile.first][cur_tile.second - 1] = false;
-            tile_deque.push_back({cur_tile.first, cur_tile.second - 1});
+        
+        else if (rank[x_ancestor] > rank[y_ancestor]) {
+            parent[y_ancestor] = x_ancestor;
         }
-        if (cur_tile.second + 1 < table[0].size() && table[cur_tile.first][cur_tile.second + 1]) {
-            table[cur_tile.first][cur_tile.second + 1] = false;
-            tile_deque.push_back({cur_tile.first, cur_tile.second + 1});
+        
+        else {
+            parent[x_ancestor] = y_ancestor;
+            rank[y_ancestor] = rank[y_ancestor] + 1;
         }
     }
-}
+};
 
 int count_islands(vector<vector<bool> > table) {
+    int max_width = 0;
+    
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
+            if (max_width < table[y].size()) {
+                max_width = table[y].size();
+            }
+        }
+    }
+    
+    Disjoint_sets islands = Disjoint_sets(table.size() * max_width);
+    vector<pair<int, int>> delta_x_y = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
+            if (table[y][x]) {
+                for (auto delta : delta_x_y) {
```

Comment:
Actually, you only need to join with 2, up and left (or up and right or similar :) ), that may make the code simpler.